File: proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthToken.java
/*
 * Copyright 2017 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.getlime.security.powerauth.sdk;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import io.getlime.security.powerauth.core.TokenCalculator;
import io.getlime.security.powerauth.exception.PowerAuthErrorCodes;
import io.getlime.security.powerauth.sdk.impl.PowerAuthPrivateTokenData;

/**
 * The <code>PowerAuthToken</code> class generates a token based authorization headers.
 * You have to use {@code PowerAuthTokenStore} to get an instance of this class.
 * <p>
 * The whole interface is thread safe.
 */
public class PowerAuthToken {

    /**
     * Reference to store which created this instance of token.
     */
    public final PowerAuthTokenStore tokenStore;
    /**
     * Token's private data
     */
    private final PowerAuthPrivateTokenData tokenData;

    /**
     * @param store {@link PowerAuthTokenStore} object who creates this token
     * @param tokenData private token data
     */
    public PowerAuthToken(@NonNull PowerAuthTokenStore store, @NonNull PowerAuthPrivateTokenData tokenData) {
        this.tokenStore = store;
        this.tokenData = tokenData;
    }

    /**
     * Return true if this token object contains a valid token data.
     * @return true if token has valid data.
     */
    public boolean isValid() {
        return tokenData.hasValidData();
    }


    /**
     * Compares this token to the specified object.
     *
     * @param anObject object to compare
     * @return true if objects are equal.
     */
    public boolean equals(Object anObject) {
        if (this == anObject) {
            return true;
        }
        if (anObject instanceof PowerAuthToken) {
            PowerAuthToken anotherToken = (PowerAuthToken) anObject;
            if (anotherToken.isValid() && this.isValid()) {
                return tokenStore == anotherToken.tokenStore &&
                        tokenData.equals(anotherToken.tokenData);
            }
        }
        return false;
    }

    /**
     * @return symbolic name of token or null if token contains an invalid data.
     */
    public @Nullable String getTokenName() {
        return tokenData.name;
    }

    /**
     * Return token's unique identifier. You normally don't need this value, but it may help
     * with application's debugging. The value identifies this token on PowerAuth server.
     *
     * @return token's unique identifier or null if token contains an invalid data.
     */
    public @Nullable String getTokenIdentifier() {
        return tokenData.identifier;
    }


    /**
     * @return true if token can generate a new header
     */
    public boolean canGenerateHeader() {
        return tokenStore.canRequestForAccessToken();
    }

    /**
     * Generates a new HTTP header for token based authorization.
     *
     * @return calculated HTTP authorization header. The header object contains an information
     *         about error, so check its <code>isValid()</code> method afterwards.
     */
    public @NonNull PowerAuthAuthorizationHttpHeader generateHeader() {
        @PowerAuthErrorCodes int errorCode;
        if (this.isValid()) {
            if (tokenStore.canRequestForAccessToken()) {
                String headerValue = TokenCalculator.calculateTokenValue(tokenData);
                if (headerValue != null) {
                    return PowerAuthAuthorizationHttpHeader.createTokenHeader(headerValue);
                } else {
                    errorCode = PowerAuthErrorCodes.PA2ErrorCodeSignatureError;
                }
            } else {
                errorCode = PowerAuthErrorCodes.PA2ErrorCodeMissingActivation;
            }
        } else {
            errorCode = PowerAuthErrorCodes.PA2ErrorCodeInvalidToken;
        }
        // In case of error, create an object with error.
        return PowerAuthAuthorizationHttpHeader.createError(errorCode);
    }
}

File: proj-xcode/PowerAuth2ForExtensions/PowerAuthToken.m
/**
 * Copyright 2021 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// PA2_SHARED_SOURCE PowerAuth2ForWatch .
// PA2_SHARED_SOURCE PowerAuth2ForExtensions .

#import <PowerAuth2ForExtensions/PowerAuthToken.h>
#import <PowerAuth2ForExtensions/PowerAuthAuthorizationHttpHeader.h>
#import <PowerAuth2ForExtensions/PowerAuthLog.h>

#import "PA2PrivateTokenInterfaces.h"
#import "PA2PrivateTokenData.h"

#if PA2_HAS_CORE_MODULE
    // Regular SDK
    @import PowerAuthCore;
#else
    // Extensions/watchOS SDK
    #import "PA2CoreCryptoUtils.h"
    #define PowerAuthCoreCryptoUtils PA2CoreCryptoUtils
#endif

@implementation PowerAuthToken
{
    PA2PrivateTokenData * _tokenData;
    __weak id<PowerAuthPrivateTokenStore> _tokenStore;
}

#pragma mark - Public getters

- (PA2PrivateTokenData*) privateTokenData
{
    return _tokenData;
}

- (NSString*) tokenName
{
    return _tokenData.name;
}

- (NSString*) tokenIdentifier
{
    return _tokenData.identifier;
}

- (BOOL) isValid
{
    return _tokenData != nil;
}

- (BOOL) canGenerateHeader
{
    return _tokenData != nil && [_tokenStore canGenerateHeaderForToken:self];
}

#pragma mark - Public methods

- (PowerAuthAuthorizationHttpHeader*) generateHeader
{
    NSData * tokenSecret = nil;
    NSString * tokenIdentifier = nil;
    
    if (!self.canGenerateHeader) {
#if defined(DEBUG)
        if (!self.isValid) {
            PowerAuthLog(@"PowerAuthToken: Token contains invalid data.");
        } else {
            PowerAuthLog(@"PowerAuthToken: The associated token store has no longer valid activation.");
        }
#endif
        return nil;
    }
    
    tokenSecret = _tokenData.secret;
    tokenIdentifier = _tokenData.identifier;

    // Prepare data for HMAC
    NSNumber * currentTimeMs = @((int64_t)([[NSDate date] timeIntervalSince1970] * 1000));
    NSString * currentTimeString = [currentTimeMs stringValue];
    NSData * currentTimeData = [currentTimeString dataUsingEncoding:NSASCIIStringEncoding];
    NSData * nonce = [PowerAuthCoreCryptoUtils randomBytes:16];
    if (nonce.length != 16) {
        PowerAuthLog(@"PowerAuthToken: Random generator did not generate enough bytes.");
        return nil;
    }
    NSMutableData * data = [nonce mutableCopy];
    [data appendBytes:"&" length:1];
    [data appendData: currentTimeData];
    // Calculate digest...
    NSData * digest = [PowerAuthCoreCryptoUtils hmacSha256:data key:tokenSecret];
    NSString * digestBase64 = [digest base64EncodedStringWithOptions:0];
    NSString * nonceBase64 = [nonce base64EncodedStringWithOptions:0];
    // Final check...
    if (digest.length == 0 || !digestBase64 || !nonceBase64 || !currentTimeString) {
        PowerAuthLog(@"PowerAuthToken: Digest calculation did fail.");
        return nil;
    }
    NSString * value = [NSString stringWithFormat:
                        @"PowerAuth version=\"3.1\""
                        @", token_id=\"%@\""
                        @", token_digest=\"%@\""
                        @", nonce=\"%@\""
                        @", timestamp=\"%@\"",
                        tokenIdentifier, digestBase64, nonceBase64, currentTimeString];
    return [PowerAuthAuthorizationHttpHeader tokenHeaderWithValue:value];
}

- (BOOL) isEqualToToken:(nonnull PowerAuthToken*)token
{
    if (token == self) {
        return YES;
    }
    if (_tokenStore != token.tokenStore || !_tokenData) {
        return NO;
    }
    return [_tokenData isEqualToTokenData:token.privateTokenData];
}

#pragma mark - Copying

- (id) copyWithZone:(NSZone *)zone
{
    return [[PowerAuthToken alloc] initWithStore:_tokenStore data:[_tokenData copy]];
}

#pragma mark - Private methods

- (id) init
{
    return nil;
}

- (id) initWithStore:(id<PowerAuthPrivateTokenStore>)store
                data:(PA2PrivateTokenData*)data
{
    self = [super init];
    if (self) {
        _tokenStore = store;
        _tokenData = data;
    }
    return self;
}

- (id<PowerAuthTokenStore>) tokenStore
{
    return _tokenStore;
}

#pragma mark - Debug

#if defined(DEBUG)
- (NSString*) description
{
    return [NSString stringWithFormat:@"<PowerAuthToken name='%@' identifier='%@' canGenerateHeader=%@>",
            self.tokenName,
            self.tokenIdentifier,
            @(self.canGenerateHeader)];
}
#endif // DEBUG

@end
